File: backend/src/main/java/dev/socialengine/config/PlatformMedia.java
package dev.socialengine.config;

import dev.socialengine.web.dto.Dtos.MediaItemDto;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.LinkedHashSet;

/**
 * Authoritative per-platform media spec (accepted formats, max size, max count).
 * Mirrors frontend/src/lib/platformMedia.ts — keep the two in sync.
 */
public final class PlatformMedia {
    private PlatformMedia() {}

    public record KindSpec(Set<String> formats, long maxSizeMB) {}
    public record Spec(KindSpec image, KindSpec video, boolean supportsAudio, int maxCount, String name) {}

    private static KindSpec k(long mb, String... fmts) {
        return new KindSpec(Set.of(fmts), mb);
    }

    public static final Map<String, Spec> SPECS = Map.of(
            "twitter", new Spec(k(5, "jpg", "jpeg", "png", "webp", "gif"), k(512, "mp4", "mov"), false, 4, "Twitter / X"),
            "instagram", new Spec(k(8, "jpg", "jpeg", "png"), k(100, "mp4", "mov"), false, 10, "Instagram"),
            "facebook", new Spec(k(30, "jpg", "jpeg", "png", "gif", "webp"), k(1024, "mp4", "mov"), false, 10, "Facebook"),
            "linkedin", new Spec(k(10, "jpg", "jpeg", "png", "gif"), k(200, "mp4"), false, 9, "LinkedIn")
    );

    /** Returns hard violations (format/size/count) for the media set across the selected platforms. */
    public static List<String> validate(List<String> platforms, List<MediaItemDto> media) {
        Set<String> errors = new LinkedHashSet<>();
        if (platforms == null || media == null || media.isEmpty()) return List.of();

        for (String id : platforms) {
            Spec spec = SPECS.get(id);
            if (spec == null) continue;

            if (media.size() > spec.maxCount()) {
                errors.add(spec.name() + " allows at most " + spec.maxCount() + " media items (you have " + media.size() + ").");
            }
            for (MediaItemDto m : media) {
                if ("audio".equals(m.type())) continue; // soft (warning handled client-side)
                KindSpec kind = "video".equals(m.type()) ? spec.video() : spec.image();
                String fmt = formatOf(m.url());
                if (!fmt.isEmpty() && !kind.formats().contains(fmt)) {
                    errors.add(fmt.toUpperCase() + " " + m.type() + " isn't supported on " + spec.name() + ".");
                }
                if (m.size() != null && m.size() > kind.maxSizeMB() * 1024L * 1024L) {
                    errors.add(("video".equals(m.type()) ? "Video" : "Image") + " exceeds " + spec.name() + "'s " + kind.maxSizeMB() + " MB limit.");
                }
            }
        }
        return List.copyOf(errors);
    }

    private static String formatOf(String url) {
        if (url == null) return "";
        String clean = url.split("[?#]")[0];
        int dot = clean.lastIndexOf('.');
        return dot >= 0 ? clean.substring(dot + 1).toLowerCase() : "";
    }
}

File: frontend/convex/lib/platformMedia.ts
// Per-platform media spec — mirror of frontend/src/lib/platformMedia.ts (pure, no React).

interface KindSpec { formats: string[]; maxSizeMB: number; }
interface Spec { image: KindSpec; video: KindSpec; supportsAudio: boolean; maxCount: number; name: string; }

const SPECS: Record<string, Spec> = {
    twitter: { image: { formats: ["jpg", "jpeg", "png", "webp", "gif"], maxSizeMB: 5 }, video: { formats: ["mp4", "mov"], maxSizeMB: 512 }, supportsAudio: false, maxCount: 4, name: "Twitter / X" },
    instagram: { image: { formats: ["jpg", "jpeg", "png"], maxSizeMB: 8 }, video: { formats: ["mp4", "mov"], maxSizeMB: 100 }, supportsAudio: false, maxCount: 10, name: "Instagram" },
    facebook: { image: { formats: ["jpg", "jpeg", "png", "gif", "webp"], maxSizeMB: 30 }, video: { formats: ["mp4", "mov"], maxSizeMB: 1024 }, supportsAudio: false, maxCount: 10, name: "Facebook" },
    linkedin: { image: { formats: ["jpg", "jpeg", "png", "gif"], maxSizeMB: 10 }, video: { formats: ["mp4"], maxSizeMB: 200 }, supportsAudio: false, maxCount: 9, name: "LinkedIn" },
};

export interface MediaItemInput { url?: string; type: string; size?: number; }

function formatOf(url: string | undefined): string {
    if (!url) return "";
    const clean = url.split(/[?#]/)[0];
    const dot = clean.lastIndexOf(".");
    return dot >= 0 ? clean.slice(dot + 1).toLowerCase() : "";
}

/** Returns hard violations (format/size/count) for the media across the selected platforms. */
export function validateMedia(media: MediaItemInput[], platforms: string[]): string[] {
    const errors = new Set<string>();
    if (!platforms || !media || media.length === 0) return [];
    for (const id of platforms) {
        const spec = SPECS[id];
        if (!spec) continue;
        if (media.length > spec.maxCount) errors.add(`${spec.name} allows at most ${spec.maxCount} media items (you have ${media.length}).`);
        for (const m of media) {
            if (m.type === "audio") continue;
            const kind = m.type === "video" ? spec.video : spec.image;
            const fmt = formatOf(m.url);
            if (fmt && !kind.formats.includes(fmt)) errors.add(`${fmt.toUpperCase()} ${m.type} isn't supported on ${spec.name}.`);
            if (m.size && m.size > kind.maxSizeMB * 1024 * 1024) errors.add(`${m.type === "video" ? "Video" : "Image"} exceeds ${spec.name}'s ${kind.maxSizeMB} MB limit.`);
        }
    }
    return [...errors];
}
